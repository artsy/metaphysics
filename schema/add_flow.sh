#!/bin/bash
# Based on http://stackoverflow.com/questions/151677/tool-for-adding-license-headers-to-source-files

for i in */*.js
do
  if ! grep -q @flow $i
  then
    (echo "" & echo "/* @flow */") > flowificator
    cat flowificator $i >$i.new && mv $i.new $i
    rm flowificator
  fi
done