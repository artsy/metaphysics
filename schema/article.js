import { has } from 'lodash';
import positron from '../lib/loaders/positron';
import cached from './fields/cached';
import AuthorType from './author';
import Image from './image';
import date from './fields/date';
import { IDFields, NodeInterface } from './object_identification';
import {
  GraphQLString,
  GraphQLObjectType,
  GraphQLNonNull,
} from 'graphql';

const ArticleType = new GraphQLObjectType({
  name: 'Article',
  interfaces: [NodeInterface],
  isTypeOf: (obj) => has(obj, 'title') && has(obj, 'author'),
  fields: () => ({
    ...IDFields,
    cached,
    author: {
      type: AuthorType,
      resolve: ({ author }) => author,
    },
    href: {
      type: GraphQLString,
      resolve: ({ slug }) => `/article/${slug}`,
    },
    published_at: date,
    slug: {
      type: GraphQLString,
    },
    thumbnail_title: {
      type: GraphQLString,
    },
    thumbnail_teaser: {
      type: GraphQLString,
    },
    thumbnail_image: {
      type: Image.type,
      resolve: ({ thumbnail_image }) => Image.resolve(thumbnail_image),
    },
    title: {
      type: GraphQLString,
    },
    updated_at: date,
  }),
});

const Article = {
  type: ArticleType,
  description: 'An Article',
  args: {
    id: {
      type: new GraphQLNonNull(GraphQLString),
      description: 'The ID of the Article',
    },
  },
  resolve: (root, { id }) => positron(`articles/${id}`),
};

export default Article;
