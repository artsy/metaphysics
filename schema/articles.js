import positron from '../lib/loaders/positron';
import Article from './article';
import ArticleSorts from './sorts/article_sorts';
import {
  GraphQLString,
  GraphQLBoolean,
  GraphQLList,
} from 'graphql';

const Articles = {
  type: new GraphQLList(Article.type),
  description: 'A list of Articles',
  args: {
    published: {
      type: GraphQLBoolean,
      defaultValue: true,
    },
    show_id: {
      type: GraphQLString,
    },
    sort: ArticleSorts,
  },
  resolve: (root, options) => {
    return positron('articles', options).then(articles => articles.results);
  },
};

export default Articles;
