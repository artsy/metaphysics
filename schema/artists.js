import gravity from '../lib/loaders/gravity';
import Artist from './artist';
import ArtistSorts from './sorts/artist_sorts';
import {
  GraphQLList,
  GraphQLInt,
} from 'graphql';

const Artists = {
  type: new GraphQLList(Artist.type),
  description: 'A list of Artists',
  args: {
    page: {
      type: GraphQLInt,
      defaultValue: 1,
    },
    size: {
      type: GraphQLInt,
    },
    sort: ArtistSorts,
  },
  resolve: (root, options) => gravity('artists', options),
};

export default Artists;
