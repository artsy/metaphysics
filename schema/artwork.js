import _ from 'lodash';
import qs from 'qs';
import cached from './fields/cached';
import Image from './image';
import Sale from './sale';
import Partner from './partner';
import Artist from './artist';
import gravity from '../lib/loaders/gravity';
import {
  GraphQLObjectType,
  GraphQLBoolean,
  GraphQLString,
  GraphQLNonNull,
  GraphQLList,
  GraphQLInt
} from 'graphql';

export let ArtworkPredicates = {
  is_contactable: (artwork, sales) => {
    return artwork.forsale && !_.isEmpty(artwork.partner) && !artwork.acquireable && !sales.length
  },
  is_in_auction: (artwork, sales) => {
    return _.any(sales, 'is_auction')
  }
}

let ArtworkType = new GraphQLObjectType({
  name: 'Artwork',
  fields: () => {

    return {
      cached: cached,
      id: {
        type: GraphQLString
      },
      href: {
        type: GraphQLString,
        resolve: (artwork) => `/artwork/${artwork.id}`
      },
      title: {
        type: GraphQLString
      },
      category: {
        type: GraphQLString
      },
      medium: {
        type: GraphQLString
      },
      date: {
        type: GraphQLString
      },
      partner:{
        type: Partner.type,
        resolve: (artwork) => {
          return artwork.partner
        }
      },
      is_contactable: {
        type: GraphQLBoolean,
        description: 'Are we able to display a contact form on artwork pages?',
        resolve: (artwork) => {
          return new Promise(resolve => {
            gravity(`related/sales`, { size: 1, active: true, artwork: [artwork.id] })
              .then(sales => {
                resolve(ArtworkPredicates.is_contactable(artwork, sales))
              });
          });
        }
      },
      is_in_auction: {
        type: GraphQLBoolean,
        description: 'Is this artwork part of an auction?',
        resolve: (artwork) => {
          return new Promise(resolve => {
            gravity(`related/sales`, { size: 1, active: true, artwork: [artwork.id] })
              .then(sales => {
                resolve(ArtworkPredicates.is_in_auction(artwork, sales))
              });
          });
        }
      },
      artist: {
        type: Artist.type,
        resolve: ({ artist }) => gravity(`artist/${artist.id}`)
      },
      dimensions: {
        type: new GraphQLObjectType({
          name: 'dimensions',
          fields: {
            in: { type: GraphQLString },
            cm: { type: GraphQLString }
          }
        })
      },
      default_image: {
        type: Image.type,
        resolve: ({ images }) => {
          let image = _.findWhere(images, { is_default: true })
          if (image === undefined) {
            image = images[0];
          }
          return image;
        }
      },
      images: {
        type: new GraphQLList(Image.type),
        args: {
          size: {
            type: GraphQLInt
          }
        },
        resolve: ({ images }, { size }) => {
          return size ? _.take(images, size) : images;
        }
      },
      sales: {
        type: new GraphQLList(Sale.type),
        args: {
          size: {
            type: GraphQLInt
          }
        },
        resolve: ({ id }, options) => gravity(`related/sales`, _.defaults(options, {
          size: 1,
          active: true,
          artwork: [id]
        }))
      }
    }
  }
});

let Artwork = {
  type: ArtworkType,
  description: 'An Artwork',
  args: {
    id: {
      type: new GraphQLNonNull(GraphQLString),
      description: 'The slug or ID of the Artwork'
    }
  },
  resolve: (root, { id }) => gravity(`artwork/${id}`)
};

export default Artwork;
