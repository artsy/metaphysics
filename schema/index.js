import Status from './status';
import Article from './article';
import Articles from './articles';
import Artwork from './artwork';
import Artworks from './artworks';
import Artist from './artist';
import Artists from './artists';
import ExternalPartner from './external_partner';
import Fair from './fair';
import Fairs from './fairs';
import Gene from './gene';
import HomePage from './home';
import OrderedSets from './ordered_sets';
import Profile from './profile';
import Partner from './partner';
import Partners from './partners';
import FilterPartners from './filter_partners';
import filterArtworks from './filter_artworks';
import PartnerCategory from './partner_category';
import PartnerCategories from './partner_categories';
import PartnerShow from './partner_show';
import PartnerShows from './partner_shows';
import Sale from './sale/index';
import Sales from './sales';
import SaleArtwork from './sale_artwork';
import Search from './search';
import Show from './show';
import TrendingArtists from './trending';
import Me from './me';
import CausalityJWT from './causality_jwt';
import Viewer from './viewer';
import ObjectIdentification from './object_identification';
import {
  GraphQLSchema,
  GraphQLObjectType,
} from 'graphql';

const schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'RootQueryType',
    fields: {
      article: Article,
      articles: Articles,
      artwork: Artwork,
      artworks: Artworks,
      artist: Artist,
      artists: Artists,
      causality_jwt: CausalityJWT,
      external_partner: ExternalPartner,
      fair: Fair,
      fairs: Fairs,
      filter_partners: FilterPartners,
      filter_artworks: filterArtworks(),
      gene: Gene,
      home_page: HomePage,
      me: Me,
      node: ObjectIdentification.NodeField,
      ordered_sets: OrderedSets,
      partner: Partner,
      partner_categories: PartnerCategories,
      partner_category: PartnerCategory,
      partner_show: PartnerShow,
      partner_shows: PartnerShows,
      partners: Partners,
      profile: Profile,
      sale: Sale,
      sale_artwork: SaleArtwork,
      sales: Sales,
      search: Search,
      show: Show,
      status: Status,
      trending_artists: TrendingArtists,
      viewer: Viewer,
    },
  }),
});

export default schema;
