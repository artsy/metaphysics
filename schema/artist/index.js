// @ts-check
import type { GraphQLFieldConfig } from "graphql"
import { pageable, getPagingParameters } from "relay-cursor-paging"
import { connectionFromArraySlice, connectionDefinitions } from "graphql-relay"
import { assign, compact, defaults, first, has, reject, includes } from "lodash"
import { exclude } from "lib/helpers"
import cached from "schema/fields/cached"
import initials from "schema/fields/initials"
import { markdown, formatMarkdownValue } from "schema/fields/markdown"
import numeral from "schema/fields/numeral"
import Image from "schema/image"
import Article from "schema/article"
import Artwork, { artworkConnection } from "schema/artwork"
import PartnerArtist from "schema/partner_artist"
import Meta from "./meta"
import PartnerShow from "schema/partner_show"
import Show from "schema/show"
import Sale from "schema/sale/index"
import ArtworkSorts from "schema/sorts/artwork_sorts"
import ArticleSorts from "schema/sorts/article_sorts"
import PartnerShowSorts from "schema/sorts/partner_show_sorts"
import SaleSorts from "schema/sale/sorts"
import ArtistCarousel from "./carousel"
import ArtistStatuses from "./statuses"
import ArtistArtworksFilters from "./artwork_filters"
import positron from "lib/loaders/legacy/positron"
import total from "lib/loaders/legacy/total"
import { GravityIDFields, NodeInterface } from "schema/object_identification"
import { GraphQLObjectType, GraphQLBoolean, GraphQLString, GraphQLNonNull, GraphQLList, GraphQLInt } from "graphql"

const artistArtworkArrayLength = (artist, filter) => {
  let length
  if (first(filter) === "for_sale") {
    length = artist.forsale_artworks_count
  } else if (first(filter) === "not_for_sale") {
    length = artist.published_artworks_count - artist.forsale_artworks_count
  } else {
    length = artist.published_artworks_count
  }
  return length
}

// TODO: Fix upstream, for now we remove shows from certain Partner types
const blacklistedPartnerTypes = ["Private Dealer", "Demo", "Private Collector", "Auction"]
const showsWithBLacklistedPartnersRemoved = shows => {
  return reject(shows, show => {
    if (show.partner) {
      return includes(blacklistedPartnerTypes, show.partner.type)
    }
    if (show.galaxy_partner_id) {
      return false
    }
    return true
  })
}

// TODO Get rid of this when we remove the deprecated PartnerShow in favour of Show.
const ShowField = {
  args: {
    active: {
      type: GraphQLBoolean,
    },
    at_a_fair: {
      type: GraphQLBoolean,
    },
    is_reference: {
      type: GraphQLBoolean,
    },
    size: {
      type: GraphQLInt,
      description: "The number of PartnerShows to return",
    },
    solo_show: {
      type: GraphQLBoolean,
    },
    status: {
      type: GraphQLString,
    },
    top_tier: {
      type: GraphQLBoolean,
    },
    visible_to_public: {
      type: GraphQLBoolean,
    },
    sort: PartnerShowSorts,
  },
  resolve: ({ id }, options, request, { rootValue: { relatedShowsLoader } }) => {
    return relatedShowsLoader(
      defaults(options, {
        artist_id: id,
        sort: "-end_at",
      })).then(shows => showsWithBLacklistedPartnersRemoved(shows))
  },
}

export const ArtistType = new GraphQLObjectType({
  name: "Artist",
  interfaces: [NodeInterface],
  isTypeOf: obj => has(obj, "birthday") && has(obj, "artworks_count"),
  fields: () => {
    return {
      ...GravityIDFields,
      cached,
      alternate_names: {
        type: new GraphQLList(GraphQLString),
      },
      articles: {
        args: {
          sort: ArticleSorts,
          limit: {
            type: GraphQLInt,
          },
        },
        type: new GraphQLList(Article.type),
        resolve: ({ _id }, options, request, { rootValue: { articlesLoader } }) =>
          articlesLoader(
            defaults(options, {
              artist_id: _id,
              published: true,
            })
          ).then(({ results }) => results),
      },
      artists: {
        type: new GraphQLList(Artist.type), // eslint-disable-line no-use-before-define
        args: {
          size: {
            type: GraphQLInt,
            description: "The number of Artists to return",
          },
          exclude_artists_without_artworks: {
            type: GraphQLBoolean,
            defaultValue: true,
          },
        },
        resolve: ({ id }, options, request, { rootValue: { relatedMainArtistsLoader } }) =>
          relatedMainArtistsLoader(
            defaults(options, {
              artist: [id],
            })
          ),
      },
      artworks: {
        type: new GraphQLList(Artwork.type),
        args: {
          size: {
            type: GraphQLInt,
            description: "The number of Artworks to return",
          },
          page: {
            type: GraphQLInt,
          },
          sort: ArtworkSorts,
          published: {
            type: GraphQLBoolean,
            defaultValue: true,
          },
          filter: {
            type: new GraphQLList(ArtistArtworksFilters),
          },
          exclude: {
            type: new GraphQLList(GraphQLString),
          },
        },
        resolve: ({ id }, options, request, { rootValue: { artistArtworksLoader } }) =>
          artistArtworksLoader(id, options).then(exclude(options.exclude, "id")),
      },
      artworks_connection: {
        type: artworkConnection,
        args: pageable({
          sort: ArtworkSorts,
          filter: {
            type: new GraphQLList(ArtistArtworksFilters),
          },
          published: {
            type: GraphQLBoolean,
            defaultValue: true,
          },
        }),
        resolve: (artist, options, request, { rootValue: { artistArtworksLoader } }) => {
          // Convert `after` cursors to page params
          const { limit: size, offset } = getPagingParameters(options)
          // Construct an object of all the params gravity will listen to
          const { sort, filter, published } = options
          const gravityArgs = { size, offset, sort, filter, published }
          return artistArtworksLoader(artist.id, gravityArgs).then(artworks =>
            connectionFromArraySlice(artworks, options, {
              arrayLength: artistArtworkArrayLength(artist, filter),
              sliceStart: offset,
            })
          )
        },
      },
      bio: {
        type: GraphQLString,
        resolve: ({ nationality, years, hometown, location }) => {
          return compact([nationality, years, hometown, location ? `based in ${location}` : undefined]).join(", ")
        },
      },
      biography: {
        type: Article.type,
        description: "The Artist biography article written by Artsy",
        resolve: ({ _id }, options, request, { rootValue: { articlesLoader } }) =>
          articlesLoader({
            published: true,
            biography_for_artist_id: _id,
            limit: 1,
          }).then(articles => first(articles.results)),
      },
      biography_blurb: {
        args: assign(
          {
            partner_bio: {
              type: GraphQLBoolean,
              description: "If true, will return featured bio over Artsy one.",
              defaultValue: false,
            },
          },
          markdown().args
        ),
        type: new GraphQLObjectType({
          name: "ArtistBlurb",
          fields: {
            credit: {
              type: GraphQLString,
              resolve: ({ credit }) => credit,
            },
            text: {
              type: GraphQLString,
              resolve: ({ text }) => text,
            },
            partner_id: {
              type: GraphQLString,
              resolve: ({ partner_id }) => partner_id,
              description: "The partner id of the partner who submitted the featured bio.",
            },
          },
        }),
        resolve: ({ blurb, id }, { format, partner_bio }, request, { rootValue: { partnerArtistsLoader } }) => {
          if (!partner_bio && blurb && blurb.length) {
            return { text: formatMarkdownValue(blurb, format) }
          }
          return partnerArtistsLoader(id, {
            size: 1,
            featured: true,
          }).then(partner_artists => {
            if (partner_artists && partner_artists.length) {
              const { biography, partner } = first(partner_artists)
              return {
                text: formatMarkdownValue(biography, format),
                credit: `Submitted by ${partner.name}`,
                partner_id: partner.id,
              }
            }
            return { text: formatMarkdownValue(blurb, format) }
          })
        },
      },
      birthday: {
        type: GraphQLString,
      },
      blurb: {
        args: markdown().args,
        type: GraphQLString,
        resolve: ({ blurb }, { format }) => {
          if (blurb.length) {
            return formatMarkdownValue(blurb, format)
          }
        },
      },
      carousel: ArtistCarousel,
      contemporary: {
        type: new GraphQLList(Artist.type), // eslint-disable-line no-use-before-define
        args: {
          size: {
            type: GraphQLInt,
            description: "The number of Artists to return",
          },
          exclude_artists_without_artworks: {
            type: GraphQLBoolean,
            defaultValue: true,
          },
        },
        resolve: ({ id }, options, request, { rootValue: { relatedContemporaryArtistsLoader } }) =>
          relatedContemporaryArtistsLoader(
            defaults(options, {
              artist: [id],
            })
          ),
      },
      consignable: {
        type: GraphQLBoolean,
        deprecationReason: "Favor `is_`-prefixed boolean attributes",
      },
      counts: {
        type: new GraphQLObjectType({
          name: "ArtistCounts",
          fields: {
            artworks: numeral(({ published_artworks_count }) => published_artworks_count),
            follows: numeral(({ follow_count }) => follow_count),
            for_sale_artworks: numeral(({ forsale_artworks_count }) => forsale_artworks_count),
            partner_shows: numeral(({ partner_shows_count }) => partner_shows_count),
            related_artists: numeral(({ id }) => total(`related/layer/main/artists`, { artist: id })),
            articles: numeral(({ _id }) =>
              positron("articles", {
                artist_id: _id,
                published: true,
                limit: 1,
                count: true,
              }).then(({ count }) => count)
            ),
          },
        }),
        resolve: artist => artist,
      },
      deathday: {
        type: GraphQLString,
      },
      display_auction_link: {
        type: GraphQLBoolean,
        deprecationReason: "Favor `is_`-prefixed boolean attributes",
      },
      exhibition_highlights: {
        args: {
          size: {
            type: GraphQLInt,
            description: "The number of Shows to return",
            defaultValue: 5,
          },
        },
        type: new GraphQLList(Show.type),
        description: "Custom-sorted list of shows for an artist, in order of significance.",
        resolve: ({ id }, options, request, { rootValue: { relatedShowsLoader } }) => {
          return relatedShowsLoader({
            artist_id: id,
            sort: "-relevance,-start_at",
            is_reference: true,
            visible_to_public: false,
            size: options.size,
          }).then(shows => showsWithBLacklistedPartnersRemoved(shows))
        },
      },
      formatted_artworks_count: {
        type: GraphQLString,
        description: "A string showing the total number of works and those for sale",
        resolve: ({ published_artworks_count, forsale_artworks_count }) => {
          let totalWorks = null
          if (published_artworks_count) {
            totalWorks = published_artworks_count + (published_artworks_count > 1 ? " works" : " work")
          }
          const forSaleWorks = forsale_artworks_count ? forsale_artworks_count + " for sale" : null
          return forSaleWorks && totalWorks ? totalWorks + ", " + forSaleWorks : totalWorks
        },
      },
      formatted_nationality_and_birthday: {
        type: GraphQLString,
        description: `A string of the form "Nationality, Birthday (or Birthday-Deathday)"`,
        resolve: ({ birthday, nationality, deathday }) => {
          let formatted_bday = !isNaN(birthday) && birthday ? "b. " + birthday : birthday
          formatted_bday = formatted_bday && formatted_bday.replace(/born/i, "b.")

          if (!isNaN(deathday) && deathday && formatted_bday) {
            formatted_bday = `${formatted_bday.replace("b. ", "")}–${deathday.match(/\d+/)}`
          }
          if (nationality && formatted_bday) {
            return nationality + ", " + formatted_bday
          }
          return nationality || formatted_bday
        },
      },
      gender: {
        type: GraphQLString,
      },
      href: {
        type: GraphQLString,
        resolve: artist => `/artist/${artist.id}`,
      },
      has_metadata: {
        type: GraphQLBoolean,
        resolve: ({ blurb, nationality, years, hometown, location }) => {
          return !!(blurb || nationality || years || hometown || location)
        },
      },
      hometown: {
        type: GraphQLString,
      },
      image: Image,
      initials: initials("name"),
      is_consignable: {
        type: GraphQLBoolean,
        resolve: ({ consignable }) => consignable,
      },
      is_display_auction_link: {
        type: GraphQLBoolean,
        description: "Only specific Artists should show a link to auction results.",
        resolve: ({ display_auction_link }) => display_auction_link,
      },
      is_followed: {
        type: GraphQLBoolean,
        resolve: ({ id }, {}, request, { rootValue: { followedArtistLoader } }) => {
          if (!followedArtistLoader) return false
          return followedArtistLoader(id).then(({ is_followed }) => is_followed)
        },
      },
      is_public: {
        type: GraphQLBoolean,
        resolve: artist => artist.public,
      },
      is_shareable: {
        type: GraphQLBoolean,
        resolve: artist => artist.published_artworks_count > 0,
      },
      location: {
        type: GraphQLString,
      },
      meta: Meta,
      nationality: {
        type: GraphQLString,
      },
      name: {
        type: GraphQLString,
      },
      partner_artists: {
        type: new GraphQLList(PartnerArtist.type),
        args: {
          size: {
            type: GraphQLInt,
            description: "The number of PartnerArtists to return",
          },
        },
        resolve: ({ id }, options, request, { rootValue: { partnerArtistsLoader } }) =>
          partnerArtistsLoader(id, options),
      },
      partner_shows: {
        type: new GraphQLList(PartnerShow.type),
        deprecationReason: "Prefer to use shows attribute",
        ...ShowField,
      },
      public: {
        type: GraphQLBoolean,
        deprecationReason: "Favor `is_`-prefixed boolean attributes",
      },
      sales: {
        type: new GraphQLList(Sale.type),
        args: {
          live: {
            type: GraphQLBoolean,
          },
          is_auction: {
            type: GraphQLBoolean,
          },
          size: {
            type: GraphQLInt,
            description: "The number of Sales to return",
          },
          sort: SaleSorts,
        },
        resolve: ({ id }, options, request, { rootValue: { relatedSalesLoader } }) =>
          relatedSalesLoader(
            defaults(options, {
              artist_id: id,
              sort: "timely_at,name",
            })
          ),
      },
      shows: {
        type: new GraphQLList(Show.type),
        ...ShowField,
      },
      sortable_id: {
        type: GraphQLString,
        description: "Use this attribute to sort by when sorting a collection of Artists",
      },
      statuses: ArtistStatuses,
      years: {
        type: GraphQLString,
      },
    }
  },
})
const Artist: GraphQLFieldConfig<ArtistType, *> = {
  type: ArtistType,
  description: "An Artist",
  args: {
    id: {
      description: "The slug or ID of the Artist",
      type: new GraphQLNonNull(GraphQLString),
    },
  },
  resolve: (root, { id }, request, resolver) => {
    if (id.length === 0) {
      return null
    }
    const { artistLoader } = (resolver.rootValue: any)
    return artistLoader(id)
  },
}
export default Artist

export const artistConnection = connectionDefinitions({
  nodeType: Artist.type,
}).connectionType
