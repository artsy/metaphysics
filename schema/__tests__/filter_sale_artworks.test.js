import schema from "schema"
import { runAuthenticatedQuery, runQuery } from "test/utils"

describe("Filter Sale Artworks", () => {
  const FilterSaleArtworks = schema.__get__("FilterSaleArtworks")
  let gravity

  beforeEach(() => {
    gravity = sinon.stub()
    gravity.with = sinon.stub().returns(gravity)
    FilterSaleArtworks.__Rewire__("gravity", gravity)
  })

  afterEach(() => {
    FilterSaleArtworks.__ResetDependency__("gravity")
  })

  it("formats the counts and aggregations, and sorts the artists correctly", () => {
    gravity
      .withArgs("filter/sale_artworks", {
        aggregations: ["total", "medium", "followed_artists", "artist"],
      })
      .returns(
        Promise.resolve({
          aggregations: {
            followed_artists: {
              value: 2,
            },
            total: {
              value: 400,
            },
            medium: {
              prints: {
                name: "Prints",
                count: 123,
              },
              painting: {
                name: "Painting",
                count: 24,
              },
            },
            artist: {
              "andy-warhol": {
                name: "Andy Warhol",
                sortable_id: "warhol-andy",
              },
              "donald-judd": {
                name: "Donald Judd",
                sortable_id: "judd-donald",
              },
              "kara-walker": {
                name: "Kara Walker",
                sortable_id: "walker-kara",
              },
            },
          },
        })
      )

    const query = `
      {
        filter_sale_artworks(
          aggregations:[TOTAL, MEDIUM, FOLLOWED_ARTISTS, ARTIST]
        ) {
          aggregations {
            slice
            counts {
              id
              name
              sortable_id
            }
          }
          counts {
            total
            followed_artists
          }
        }
      }
    `

    return runQuery(query).then(({ filter_sale_artworks: { aggregations, counts } }) => {
      expect(counts).toEqual({ followed_artists: 2, total: 400 })
      expect(aggregations).toEqual([
        {
          counts: [
            { id: "prints", name: "Prints", sortable_id: null },
            { id: "painting", name: "Painting", sortable_id: null },
          ],
          slice: "MEDIUM",
        },
        {
          counts: [
            {
              id: "donald-judd",
              name: "Donald Judd",
              sortable_id: "judd-donald",
            },
            {
              id: "kara-walker",
              name: "Kara Walker",
              sortable_id: "walker-kara",
            },
            {
              id: "andy-warhol",
              name: "Andy Warhol",
              sortable_id: "warhol-andy",
            },
          ],
          slice: "ARTIST",
        },
      ])
    })
  })

  describe("sale_artworks_connection", () => {
    it("filters artworks based on aggregations and returns pageable collection", () => {
      gravity
        .withArgs("filter/sale_artworks", {
          aggregations: ["total", "followed_artists"],
          include_artworks_by_followed_artists: true,
          size: 20,
        })
        .returns(
          Promise.resolve({
            aggregations: {
              followed_artists: {
                value: 3,
              },
              total: {
                value: 3,
              },
            },
          })
        )

      // Paginate
      gravity
        .withArgs("filter/sale_artworks", {
          page: 1,
          size: 3,
          offset: 0,
        })
        .returns(
          Promise.resolve({
            hits: [
              {
                id: "christopher-anderson-reflection-in-window-in-altamira-caracas-venezuela",
              },
              {
                id: "eduardo-abaroa-insercion-arqueologica-285-1",
              },
              {
                id: "francis-alys-the-modern-procession",
              },
            ],
          })
        )

      const query = `
        {
          filter_sale_artworks(
            aggregations:[TOTAL, FOLLOWED_ARTISTS],
            include_artworks_by_followed_artists: true,
            size: 20
          ) {
            counts {
              total
              followed_artists
            }
            sale_artworks_connection(first: 3) {
              pageInfo {
                hasNextPage
              }
              edges {
                cursor
                node {
                  id
                }
              }
            }
          }
        }
      `

      return runAuthenticatedQuery(query).then(data => {
        const { filter_sale_artworks: { sale_artworks_connection: { pageInfo, edges } } } = data
        expect(data).toMatchSnapshot()
        expect(pageInfo.hasNextPage).toEqual(false)
        expect(edges.length).toEqual(3)
      })
    })
  })
})
