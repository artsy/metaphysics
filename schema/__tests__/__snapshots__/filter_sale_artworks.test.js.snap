// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Sale Artworks sale_artworks_connection filters artworks based on aggregations and returns pageable collection 1`] = `
Object {
  "filter_sale_artworks": Object {
    "counts": Object {
      "followed_artists": 3,
      "total": 3,
    },
    "sale_artworks_connection": Object {
      "edges": Array [
        Object {
          "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
          "node": Object {
            "id": "christopher-anderson-reflection-in-window-in-altamira-caracas-venezuela",
          },
        },
        Object {
          "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
          "node": Object {
            "id": "eduardo-abaroa-insercion-arqueologica-285-1",
          },
        },
        Object {
          "cursor": "YXJyYXljb25uZWN0aW9uOjI=",
          "node": Object {
            "id": "francis-alys-the-modern-procession",
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": false,
      },
    },
  },
}
`;
